protocol/strict-derivation.md

```diff
@@ -0,0 +1,269 @@
+  - Option 2: discard the foreign frame and wait for the duration of the channel timeout for the
+    next correct frame and only then discard the current staging channel and open a new staging
+    channel for the next first frame.
+- How do we handle the current staging channel if an out of order frame (that is not the first
+frame) arrives? There's only one valid option, to drop the out-of-order frame, keeping the currently
+staged channel. Note that this will give the staging channel a chance to be continued or closed
+properly by the next in-order frame. If this doesn't happen, at the lastest the next first frame
+will drop the staging channel and start a new one.
+
+### Proposed solution
+
+Applied to above open question, this means choosing option 1. This has the
+additional advantage that the sync start algorithm simplifies significantly, because a fist frame
+_always_ marks the start of a new channel, and guarantees that there won't be any buffered channels
+in the channel bank.
+
+## Out of order batches
+
+The batch queue also becomes simpler as batches have to arrive in order. 
+
+- How to treat _older_ batches is easy: they are dropped.
+- Similarly to out of order frames, how to treat future batches (gaps) has two options:
+  - Option 1: immediately fill the gap with deposit-only blocks, and then process the batch.
+  - Option 2: drop future batches, and wait for up to a full sequencing window for the next
+  batch to arrive.
+
+### Proposed solution
+
+Following the principle of fast derivation, option 1 is proposed to be chosen. Besides guaranteeing
+faster derivation, it also avoids waiting for full sequencing windows on gaps to be resolved, which
+is quite unlikely, unless the batcher implementation would be specially tuned to backfill gaps. If a
+gap occurs, it is actually much more likely that the next batches will be even further in the
+future, so option 2 would realistically almost always be a bad choice. Plus, the sync start
+algorithm benefits from option 1 as well, as it leads to faster resolution of uncertainty caused
+by gaps.
+
+It should be noted that if the blocks after the gap actually come from a chain where the gaps
+aren't deposit-only blocks, then these blocks would have mismatched parent hashes and be likely
+discarded. This is intentional. This design allows for not having to submit empty batches to
+derive deposit-only blocks, but instead just pick the next non-deposit-only block.
+
+Writing this, I realize that this mechanism could even be used to encode large spans of empty
+batches, as long as the sequencer is creating unsafe blocks that follow the same L1 origins as the
+auto-derivation would for gaps. However, this would need to be investigated more deeply.
+
+Note that the new strict ordering rules of the batch queue will always lead to an empty batch queue
+when the origin of the derivation pipeline progresses to the next L1 block (what about
+`undecided` though?).
```

Comment:
I am not familiar with `undecided`, and I see another reference to it below. Can you explain how these behave briefly?

Reply:
The `undecided` status is still not 💯 clear to me, which is why I added those questions in brackets. What I understand from the implementation (files `batches.go` and `batch_queue.go`) is that if there's missing L1 or L2 data (e.g. due to temporary broken connections?), the batch is re-queued for later checking and no batch is processed at this point. I think we will just retain this behavior.

The difference to `future` is that a future batch is already determined to be out of order and lies in the future. The `future` case (a gap) will be treated differently in Holocene, and the gap will immediately be derived as deposit-only blocks. It is noteworthy that the future batch can then only be valid if it built on top of a gap of empty batches.

Added this to the design doc.